lightning-payjoin/src/lib.rs

```diff
@@ -1 +1,316 @@
+	}
+
+	fn identify_my_outputs(&self, script: &ScriptBuf) -> Result<bool, Error> {
+		self.receiver_handler.identify_my_outputs(script)
+	}
+
+	fn can_broadcast(&self, transaction: &Transaction) -> Result<bool, Error> {
+		self.receiver_handler.can_broadcast(transaction)
+	}
+
+	fn sign_tx(&self, psbt: &Psbt) -> Result<Psbt, Error> {
+		self.receiver_handler.sign_tx(psbt)
+	}
+
+	async fn extract_request_parts(
+		&self, request: Request<impl Body>,
+	) -> Result<(Vec<u8>, String, RequestHeaders), Error> {
+		let (parts, body) = request.into_parts();
+		let body = match body.collect().await {
+			Ok(body) => body.to_bytes().to_vec(),
+			Err(_e) => panic!("panicesh"),
+		};
+		let query = parts.uri.query().unwrap_or("").to_string();
+		let headers = parts.headers;
+		let headers = RequestHeaders::from(headers);
+		Ok((body, query, headers))
+	}
+
+	fn validate_payjoin_request(
+		&self, body: &[u8], query: &str, headers: RequestHeaders, min_fee_rate: Option<FeeRate>,
+	) -> Result<ProvisionalProposal, Error> {
+		let proposal = payjoin::receive::UncheckedProposal::from_request(body, query, headers)?;
+		let proposal =
+			proposal.check_broadcast_suitability(min_fee_rate, |t| Ok(self.can_broadcast(t)?))?;
+		let proposal = proposal.check_inputs_not_owned(|script| {
+			Ok(self.check_inputs_not_owned(&script.to_owned())?)
+		})?;
+		let proposal = proposal.check_no_mixed_input_scripts()?;
+		let proposal = proposal.check_no_inputs_seen_before(|_outpoint| Ok(false))?;
+		let original_proposal = proposal.clone().identify_receiver_outputs(|script| {
+			Ok(self.identify_my_outputs(&script.to_owned())?)
+		})?;
+		Ok(original_proposal)
+	}
+
+	async fn get_next_channel(&self, payjoin_amount: u64) -> Option<scheduler::ScheduledChannel> {
+		self.channel_scheduler.lock().await.get_next_channel(payjoin_amount)
+	}
+
+	fn accept_payjoin_without_channel_opening(
+		&self, provisional_proposal: ProvisionalProposal,
+	) -> Result<String, Error> {
+		// fixme: instead of psbt.clone() we should sign the psbt,
+		// but BDK adds bip32 derivation paths to the psbt which is not supported by payjoin
+		let finalized_proposal =
+			provisional_proposal.finalize_proposal(|psbt| Ok(psbt.clone()), None)?;
+		Ok(finalized_proposal.psbt().to_string())
+	}
+
+	async fn handle_incoming_payjoin_request(
+		&self, request: Request<impl Body>,
+	) -> Result<String, Error> {
+		let (body, query, headers) = self.extract_request_parts(request).await?;
+		let provisional_proposal =
+			self.validate_payjoin_request(body.as_slice(), &query, headers.clone(), None)?;
+		let payjoin_amount = provisional_proposal.payjoin_amount();
+		if let Some(channel) = self.get_next_channel(payjoin_amount).await {
+			let mut proposal = provisional_proposal.clone();
+			let output_script = channel.output_script().expect("Output script should exist");
+			// substitute the output script with the channel output script
+			proposal.substitute_output_script(output_script);
```

Comment:
https://github.com/payjoin/rust-payjoin/pull/233/commits/28533dabe712d07bc1454739a832efcb5f2c9ff4#diff-5fba6fc39c59b07e5980d87da92537138068364d192a2b63eee5e57af01a07bbR476

Reply:
this is where we change output of the psbt to point to the funding tx multisig